### ISSUE TITLE
njit on f2py wrapped code

### ISSUE THREAD

**Issue by Mobbybobby** [numba/numba#7550]
---

## Feature request

Hi,

is there a way to use numba in no python mode on functions that were wrapped with f2py? 

thanks

**Comment by stuartarchibald** [comment]
@Mobbybobby Numba has support for calling functions from external libraries via `ctypes` (docs: https://numba.readthedocs.io/en/stable/reference/pysupported.html?highlight=ctypes#ctypes). Should the `f2py` compiled library export a symbol to the underlying fortran function then it should be possible to bind to this. Note that extreme care will be needed with regards to memory managements (i.e. be aware of returning allocations from Fortran and potential lifetime issues).

**Comment by stuartarchibald** [comment]
xref: https://github.com/numba/numba/issues/3086 which should provide some hints and examples.

**Comment by Mobbybobby** [comment]
Thanks for the response. When trying to get the adress I get this error:
`AttributeError: module 'BPS' has no attribute '__pyx_capi__'`

Sry.. I'm not familiar with ctypes :/

**Comment by Mobbybobby** [comment]
I got the wrapping to work but numba doesn't work:
```
Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'c_double' of type Module(<module 'ctypes' from 'C:\\Program Files (x86)\\Python37-32\\lib\\ctypes\\__init__.py'>)
```

**Comment by Mobbybobby** [comment]
This is the function that produces that error:
```
@njit(error_model="numpy")
def dedx_BPS(E, T, n, m_i, z_b):  # z_b : Projectile charge, E : Energy in MeV, T : Temperature in keV,
    # n: Electron density in 1/cc, m_i: Projectile mass in g, Output in MeV/mm
    c_double = ct.c_double
    pointer = ct.POINTER(c_double)
    byref = ct.byref
    lib = ct.cdll.LoadLibrary('lib.so')
    dedx_BPS1 = lib.dedx_bps_
    dedx_BPS1.argtypes = [ct.POINTER(ct.c_int),
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer,
                          pointer]
    ne = np.array([n], dtype=c_double, order='F')
    Te = np.array([1 / T], dtype=c_double, order='F')
    zb = np.array([-1], dtype=c_double, order='F')
    mkev = constants.c ** 2 / constants.eV / 1000
    mb = np.array([constants.m_e * mkev], dtype=ct.c_double, order='F')
    x1 = c_double(0)
    x2 = c_double(0)
    x3 = c_double(0)
    x4 = c_double(0)
    x5 = c_double(0)
    x6 = c_double(0)
    x7 = c_double(0)
    x8 = c_double(0)
    x9 = c_double(0)
    dedx_BPS1(byref(ct.c_int(1)), byref(c_double(1e-7)), byref(c_double(E * 1000)),
              byref(c_double(z_b)), byref(c_double(m_i / 1000 * mkev)),
              Te.ctypes.data_as(pointer), zb.ctypes.data_as(pointer),
              mb.ctypes.data_as(pointer), ne.ctypes.data_as(pointer),
              byref(x1), byref(x2), byref(x3), byref(x4), byref(x5), byref(x6),
              byref(x7), byref(x8), byref(x9))
    return np.ctypeslib.as_array(x1)
```

It seems that numba doesn't know ctypes?
Pls help

**Comment by Mobbybobby** [comment]
I'm on windows and python 3.7 btw.

**Comment by Mobbybobby** [comment]
@esc Can you help me? Should i start a new issue?

**Comment by stuartarchibald** [comment]
@Mobbybobby the Numba engineers are currently preparing a release candidate for Numba 0.55 https://github.com/numba/numba/milestone/56 and so are unfortunately a little pushed for time at present. I'll help you with this if you are still stuck once the RC is done (so hopefully in about a week), alternatively you could ask on https://numba.discourse.group/ where the Numba community might be able to provide support more quickly (it's also a better place for asking questions than the issue tracker). Thanks for your understanding.

Side note: Just skimming the code above, Numba can understand the ctypes _types_ but not the entire module. You need to get your code to look something more like: https://numba.discourse.group/t/how-to-register-an-existing-builtin-function/350/4

**Comment by Mobbybobby** [comment]
@stuartarchibald Thank you for responding. I'll try my best getting it to work with your example. I will report about my success in a week :)

**Comment by Mobbybobby** [comment]
hey @stuartarchibald.

I think i got it working like this:

```
import numpy as np

import scipy.constants as constants

import ctypes as ct

from numba import njit

lib = ct.cdll.LoadLibrary('lib.so')
dedx_BPS1 = lib.dedx_bps_
c_void_p = ct.c_void_p
dedx_BPS1.argtypes = [c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p]


@njit(error_model="numpy")
def dedx_BPS(E, T, n, m_i, z_b):  # z_b : Projectile charge, E : Energy in MeV, T : Temperature in keV,
    # n: Electron density in 1/cc, m_i: Projectile mass in g, Output in MeV/mm
    float64 = np.float64
    array = np.array
    ne = array([n], dtype=float64)
    Te = array([1 / T], dtype=float64)
    zb = array([-1], dtype=float64)
    mkev = constants.c ** 2 / constants.eV / 1000
    mb = array([constants.m_e * mkev], dtype=float64)
    scale = array(1e-7, dtype=float64)
    E = array(E * 1000, dtype=float64)
    z_b = array(z_b, dtype=float64)
    m_i = array(m_i / 1000 * mkev, dtype=float64)
    x1 = array(0.0)
    x2 = array(0.0)
    x3 = array(0.0)
    x4 = array(0.0)
    x5 = array(0.0)
    x6 = array(0.0)
    x7 = array(0.0)
    x8 = array(0.0)
    x9 = array(0.0)
    dedx_BPS1(array(1).ctypes.data, scale.ctypes.data, E.ctypes.data,
              z_b.ctypes.data, m_i.ctypes.data,
              Te.ctypes.data, zb.ctypes.data,
              mb.ctypes.data, ne.ctypes.data,
              x1.ctypes.data, x2.ctypes.data, x3.ctypes.data, x4.ctypes.data, x5.ctypes.data, x6.ctypes.data,
              x7.ctypes.data, x8.ctypes.data, x9.ctypes.data)
    return x1 * 10 ** 3

```

But it is slower than using f2py without njit. Did I do something wrong?

**Comment by Mobbybobby** [comment]
@stuartarchibald I got working like this:

```
import numpy as np

import scipy.constants as constants

import ctypes as ct

from numba import njit

lib = ct.cdll.LoadLibrary('lib.so')
dedx_BPS1 = lib.dedx_bps_
c_void_p = ct.c_void_p
dedx_BPS1.argtypes = [c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p,
                      c_void_p]


@njit(error_model="numpy")
def dedx_BPS(E, T, n, m_i, z_b):  # z_b : Projectile charge, E : Energy in MeV, T : Temperature in keV,
    # n: Electron density in 1/cc, m_i: Projectile mass in g, Output in MeV/mm
    float64 = np.float64
    array = np.array
    ne = array([n], dtype=float64)
    Te = array([1 / T], dtype=float64)
    zb = array([-1], dtype=float64)
    mkev = constants.c ** 2 / constants.eV / 1000
    mb = array([constants.m_e * mkev], dtype=float64)
    scale = array(1e-7, dtype=float64)
    E = array(E * 1000, dtype=float64)
    z_b = array(z_b, dtype=float64)
    m_i = array(m_i / 1000 * mkev, dtype=float64)
    x1 = array(0.0)
    x2 = array(0.0)
    x3 = array(0.0)
    x4 = array(0.0)
    x5 = array(0.0)
    x6 = array(0.0)
    x7 = array(0.0)
    x8 = array(0.0)
    x9 = array(0.0)
    dedx_BPS1(array(0).ctypes.data, scale.ctypes.data, E.ctypes.data,
              z_b.ctypes.data, m_i.ctypes.data,
              Te.ctypes.data, zb.ctypes.data,
              mb.ctypes.data, ne.ctypes.data,
              x1.ctypes.data, x2.ctypes.data, x3.ctypes.data, x4.ctypes.data, x5.ctypes.data, x6.ctypes.data,
              x7.ctypes.data, x8.ctypes.data, x9.ctypes.data)
    return x1 * 10 ** 3
```

But it is slower than using f2py to wrap without njit:

```
import numpy as np

import scipy.constants as constants

import BPS


def dedx_BPS(E, T, n, m_i, z_b):  # z_b : Projectile charge, E : Energy in MeV, T : Temperature in keV,
    # n: Electron density in 1/cc, m_i: Projectile mass in g, Output in MeV/mm
    float64 = np.float64
    mkev = constants.c ** 2 / constants.eV / 1000

    return \
        BPS.dedx_bps(nni=int(0), scale=float64(1e-7), ep=float64(E * 1000), zp=float64(z_b),
                     mp=float64(m_i / 1000 * mkev),
                     betab=float64(1 / T), zb=float64(-1), mb=float64(constants.m_e * mkev), nb=float64(n))[0] * 10 ** 3
```
I tested it like this:
```

import scipy.constants as constants

import _BPS  # f2py
import __BPS  # ctypes

from time import time

E = 8
m_i = 12 * constants.u * 1000
z_b = 5
T = 0.2
n = 1e21
t = time()
for i in range(1, 1000):
    rpa = _BPS.dedx_BPS(E, T, n, m_i, z_b)
t1 = time()
for i in range(1, 1000):
    cas = __BPS.dedx_BPS(E, T, n, m_i, z_b)
t2 = time()
print(rpa, cas)
print(t1-t)
print(t2-t1)
```
Output:
```
3.4902877171282793 3.4902877171282793
4.714770793914795
5.245108366012573
```

Is there anything I can do to improve speed?

**Comment by stuartarchibald** [comment]
@Mobbybobby Great that you got something working! The above looks approximately as I'd expect it would as it follows the `ctypes` support. Note that you may need to use a `sink` function after the call to the `f2py` function, the purposes of this is to keep the variables used as arguments "alive" (Numba cannot at present work out they are still live so might "delete" them). This would look something like:
```python
@njit
def sink(*args):  # does nothing
  return args[0]

@njit
def foo():
   a = np.array(0.0)
   ... # more allocations for args
   f2py_function_call(a.ctypes.data, ... #more args)
   sink(a, ... # more args)
```
See also the Numba internal version of this, definition:
https://github.com/numba/numba/blob/4cc738b0cea246bb0cfd9807a8c783be0fa5ddac/numba/np/linalg.py#L826-L829
example use:
https://github.com/numba/numba/blob/4cc738b0cea246bb0cfd9807a8c783be0fa5ddac/numba/np/linalg.py#L865-L866

RE:
> Is there anything I can do to improve speed?

I wouldn't have thought so, Numba cannot speed up functions it's not compiling. The performance of the `f2py` function is largely dependent on the way it was compiled etc. There's also obviously going to be overheads associated with the call in Numba such as allocating temporary arrays for pass-by-reference variables. Were this routine used as part of a larger algorithm, say inside a loop, then these allocations could be hoisted out of the loop and their cost somewhat amoritised.

**Comment by Mobbybobby** [comment]
Thank you. This helped a lot!
